taf/GitRepository.py

```diff
@@ -20,32 +22,85 @@ def __init__(self, root_dir, target_path=None, repo_urls=None, additional_info=N
     root_dir.
     """
     self.target_path = target_path
-    self.repo_path = str((Path(root_dir) / (target_path or '')).resolve(True))
+    self.repo_path = str((Path(root_dir) / (target_path or '')).resolve())
     self.repo_urls = repo_urls
     self.additional_info = additional_info
+    self.bare = bare
+
+  @property
+  def is_git_repository(self):
+    return self._is_git_repository()
 
   @property
   def name(self):
     return os.path.basename(self.repo_path)
 
-  def _git(self, cmd):
-    return run(f'git -C {self.repo_path} {cmd}')
+  def _git(self, cmd, *args):
+    """Call git commands in subprocess
+
+    E.G.:
+      self._git('checkout {}', branch_name)
+    """
+    if len(args):
+      cmd = cmd.format(*args)
+    return run('git -C {} {}'.format(self.repo_path, cmd))
+
+  def all_commits_since_commit(self, since_commit):
+    if since_commit is not None:
+      commits = self._git('rev-list {}..HEAD', since_commit).strip()
+    else:
+      commits = self._git('log --format=format:%H').strip()
+    if not commits:
+      return []
+    commits = commits.split('\n')
+    commits.reverse()
+    return commits
+
+  def all_fetched_commits(self, branch='master'):
+    commits = self._git('rev-list ..origin/{}'.format(branch)).strip()
+    if not commits:
+      return []
+    commits = commits.split('\n')
+    commits.reverse()
+    return commits
 
   def checkout_branch(self, branch_name, create=False):
     """Check out the specified branch. If it does not exists and
     the create parameter is set to True, create a new branch.
     If the branch does not exist and create is set to False,
     raise an exception."""
     try:
-      self._git(f'checkout {branch_name}')
+      self._git('checkout {}', branch_name)
     except subprocess.CalledProcessError as e:
       if create:
         self.create_and_checkout_branch(branch_name)
       else:
         raise(e)
 
+  def clone(self, no_checkout=False, from_filesystem=True):
+    # TODO this creates something like smc-law/smc-law instead of just smc-law
+    shutil.rmtree(self.repo_path, True)
```

Comment:
Do you plan to remove this now or to add another parameter to indicate that existing repo will be removed from local and cloned again?

Reply:
I would like to integrate the updater with the publish server before merging, as there will surely be many more changes. Will go through the code and clean it up before this is merging.

Reply:
Let's leave this for now. Clone will not be called if the repository already exists. Let's leave the TODO as well and give it another thought later.